crates/node/src/command.rs

```diff
@@ -36,17 +36,33 @@ impl SubstrateCli for Cli {
 
     fn load_spec(&self, id: &str) -> Result<Box<dyn sc_service::ChainSpec>, String> {
         Ok(match id {
-            "dev" => {
-                let enable_manual_seal = self.sealing.map(|_| true);
-                Box::new(chain_spec::development_config(
-                    enable_manual_seal,
-                    self.madara_path.clone().expect("`madara_path` expected to be set with clap default value"),
-                )?)
+            DEV_CHAIN_ID => {
+                let enable_manual_seal = self.run.sealing.map(|_| true);
+                let base_path = self
+                    .run
+                    .base
+                    .shared_params
+                    .base_path
+                    .as_ref()
+                    .map(BasePath::new)
+                    .unwrap_or_else(|| BasePath::from_project("", "", &<Cli as SubstrateCli>::executable_name()));
+                Box::new(chain_spec::development_config(enable_manual_seal, base_path)?)
             }
-            "" | "local" | "madara-local" => Box::new(chain_spec::local_testnet_config(
-                self.madara_path.clone().expect("`madara_path` expected to be set with clap default value"),
+            SHARINGAN_CHAIN_ID => Box::new(chain_spec::ChainSpec::from_json_bytes(
+                &include_bytes!("../../../configs/chain-specs/testnet-sharingan-raw.json")[..],
             )?),
-            path => Box::new(chain_spec::ChainSpec::from_json_file(std::path::PathBuf::from(path))?),
+            "" | "local" | "madara-local" => {
+                let base_path = self
+                    .run
+                    .base
+                    .shared_params
+                    .base_path
+                    .as_ref()
+                    .map(BasePath::new)
+                    .unwrap_or_else(|| BasePath::from_project("", "", &<Cli as SubstrateCli>::executable_name()));
```

Comment:
In order to avoid code duplication?

Just the `self
                    .run
                    .base
                    .shared_params
                    .base_path
                    .as_ref()
                    .map(BasePath::new)` part then. Not the `.unwrap_or_else(|| BasePath::from_project("", "", &<Cli as SubstrateCli>::executable_name()))` which is aware of `<Cli as SubstrateCli>`.

Reply:
Oh, yes, you are right `ExtendedRunCmd` is unaware of the executable name. So, not the best place for the method. Should go to `Cli` then, or left as it is. But, I personally don't like it, 8 duplicated lines of code one after another.

Reply:
btw, `executable_name` is not a method. you can still use it
```
impl ExtendedRunCmd {
    fn base_path(&self) -> Result<BasePath> {
        self.base.base_path().map(|maybe_path| {
            maybe_path.unwrap_or_else(|| BasePath::from_project("", "", &<Cli as SubstrateCli>::executable_name()))
        })
    }
}
```

Reply:
done